File: nginx-lua/lua/metrics.lua
-- metrics.lua — Prometheus metrics rendering.

local _M = {}

--- Render Prometheus text-format metrics from the file registry.
-- @param entries  array of file entry tables from registry:list_all()
-- @param duplicates  number of duplicate file keys
function _M.render(entries, duplicates)
    local lines = {}

    -- Volume files by bucket
    local bucket_counts = {}
    local week_count  = 0
    local month_count = 0
    local now = os.time()
    local week_ago  = now - (7  * 24 * 3600)
    local month_ago = now - (31 * 24 * 3600)

    for _, entry in ipairs(entries) do
        local bkt = entry.bucket or "unknown"
        bucket_counts[bkt] = (bucket_counts[bkt] or 0) + 1

        -- Parse ISO 8601 timestamp for upload window metrics
        local lm = entry.last_modified or ""
        if lm ~= "" then
            -- Try parsing "2026-05-05T10:00:00Z" or similar
            local y, mo, d, h, mi, s = lm:match("(%d+)-(%d+)-(%d+)T(%d+):(%d+):(%d+)")
            if y then
                local t = os.time({
                    year = tonumber(y), month = tonumber(mo), day = tonumber(d),
                    hour = tonumber(h), min = tonumber(mi), sec = tonumber(s),
                })
                if t and t >= week_ago and t <= now then
                    week_count = week_count + 1
                end
                if t and t >= month_ago and t <= now then
                    month_count = month_count + 1
                end
            end
        end
    end

    -- parparchik_volume_files
    lines[#lines + 1] = "# HELP parparchik_volume_files Current number of known files by bucket."
    lines[#lines + 1] = "# TYPE parparchik_volume_files gauge"
    -- Sort bucket names for deterministic output
    local sorted_buckets = {}
    for bkt in pairs(bucket_counts) do
        sorted_buckets[#sorted_buckets + 1] = bkt
    end
    table.sort(sorted_buckets)
    for _, bkt in ipairs(sorted_buckets) do
        lines[#lines + 1] = string.format('parparchik_volume_files{volume="%s"} %d', bkt, bucket_counts[bkt])
    end

    -- parparchik_duplicate_files
    lines[#lines + 1] = "# HELP parparchik_duplicate_files Number of file keys that exist in more than one S3 bucket."
    lines[#lines + 1] = "# TYPE parparchik_duplicate_files gauge"
    lines[#lines + 1] = string.format("parparchik_duplicate_files %d", duplicates or 0)

    -- parparchik_uploads_per_week
    lines[#lines + 1] = "# HELP parparchik_uploads_per_week Known uploaded file versions modified during the last 7 days."
    lines[#lines + 1] = "# TYPE parparchik_uploads_per_week gauge"
    lines[#lines + 1] = string.format("parparchik_uploads_per_week %d", week_count)

    -- parparchik_uploads_per_month
    lines[#lines + 1] = "# HELP parparchik_uploads_per_month Known uploaded file versions modified during the last 31 days."
    lines[#lines + 1] = "# TYPE parparchik_uploads_per_month gauge"
    lines[#lines + 1] = string.format("parparchik_uploads_per_month %d", month_count)

    return table.concat(lines, "\n") .. "\n"
end

return _M

File: golang/internal/metricsapi/metrics.go
// Package metricsapi renders parparchik's Prometheus metrics from the
// current catalog state, using the same metric names as the Lua
// implementation's hand-written text renderer.
package metricsapi

import (
	"net/http"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"

	"github.com/rachlenko/parparchik/golang/internal/catalog"
)

const (
	week  = 7 * 24 * time.Hour
	month = 31 * 24 * time.Hour
)

// Metrics holds the registered Prometheus collectors.
type Metrics struct {
	registry        *prometheus.Registry
	volumeFiles     *prometheus.GaugeVec
	duplicateFiles  prometheus.Gauge
	uploadsPerWeek  prometheus.Gauge
	uploadsPerMonth prometheus.Gauge
}

// New creates a Metrics instance with its own registry, so tests can create
// isolated instances without colliding on the global default registry.
func New() *Metrics {
	m := &Metrics{
		registry: prometheus.NewRegistry(),
		volumeFiles: prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Name: "parparchik_volume_files",
			Help: "Current number of known files by bucket.",
		}, []string{"volume"}),
		duplicateFiles: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "parparchik_duplicate_files",
			Help: "Number of file keys that exist in more than one S3 bucket.",
		}),
		uploadsPerWeek: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "parparchik_uploads_per_week",
			Help: "Known uploaded file versions modified during the last 7 days.",
		}),
		uploadsPerMonth: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "parparchik_uploads_per_month",
			Help: "Known uploaded file versions modified during the last 31 days.",
		}),
	}
	m.registry.MustRegister(m.volumeFiles, m.duplicateFiles, m.uploadsPerWeek, m.uploadsPerMonth)
	return m
}

// Update recomputes every gauge from the given catalog entries and duplicate
// count.
func (m *Metrics) Update(entries []catalog.Entry, duplicates int) {
	m.volumeFiles.Reset()

	bucketCounts := make(map[string]float64)
	now := time.Now()
	var weekCount, monthCount float64

	for _, entry := range entries {
		bucket := entry.Bucket
		if bucket == "" {
			bucket = "unknown"
		}
		bucketCounts[bucket]++

		if t, ok := parseLastModified(entry.LastModified); ok {
			age := now.Sub(t)
			if age >= 0 && age <= week {
				weekCount++
			}
			if age >= 0 && age <= month {
				monthCount++
			}
		}
	}

	for bucket, count := range bucketCounts {
		m.volumeFiles.WithLabelValues(bucket).Set(count)
	}
	m.duplicateFiles.Set(float64(duplicates))
	m.uploadsPerWeek.Set(weekCount)
	m.uploadsPerMonth.Set(monthCount)
}

// Handler returns the http.Handler that serves this Metrics' Prometheus
// text-format exposition.
func (m *Metrics) Handler() http.Handler {
	return promhttp.HandlerFor(m.registry, promhttp.HandlerOpts{})
}

// parseLastModified accepts RFC3339 timestamps (what objectstore emits) as
// well as the plain "YYYY-MM-DDTHH:MM:SSZ" form the Lua service wrote.
func parseLastModified(s string) (time.Time, bool) {
	if s == "" {
		return time.Time{}, false
	}
	if t, err := time.Parse(time.RFC3339, s); err == nil {
		return t, true
	}
	return time.Time{}, false
}
